src/app/zones/data/index.ts

```diff
@@ -1,36 +1,89 @@
-import { getIsConnected } from '@/app/subscriptions/data'
-import type { ZoneOverview } from '@/types/index.d'
+import type { PaginatedApiListResponse as CollectionResponse } from '@/types/api.d'
+import type {
+  ZoneOverview as PartialZoneOverview,
+  ZoneInsight as PartialZoneInsight,
+  Zone as PartialZone,
+  KDSSubscription,
+} from '@/types/index.d'
 import { get } from '@/utilities/get'
 
-// TODO(jc): These two are extremely similar, see if we can merge
-export function getZoneControlPlaneEnvironment(zoneOverview: ZoneOverview): string {
-  // TODO(jc): is it ok to use the config from the first subscription we find here?
-  for (const subscription of zoneOverview.zoneInsight?.subscriptions ?? []) {
-    if (subscription.config) {
-      return JSON.parse(subscription.config).environment
-    }
-  }
-  return ''
+export type Zone = PartialZone & {
+  enabled: boolean
+}
+export type ZoneInsight = PartialZoneInsight & {
+  connectedSubscription?: KDSSubscription
+  config: Record<string, unknown>
+  authenticationType: string
+  environment: string
+}
+export type ZoneOverview = PartialZoneOverview & {
+  zoneInsight?: ZoneInsight
+  zone: Zone
+  state: 'online' | 'offline' | 'disabled'
 }
-export function getZoneDpServerAuthType(zone: ZoneOverview): string {
-  const subscriptions = zone.zoneInsight?.subscriptions ?? []
-  if (subscriptions.length > 0) {
-    const lastSubscription = subscriptions[subscriptions.length - 1]
-    if (lastSubscription.config) {
-      const parsedConfig = JSON.parse(lastSubscription.config)
-      return get(parsedConfig, 'dpServer.auth.type', '')
+
+export const Zone = {
+  fromObject: (item: PartialZone): Zone => {
+    return {
+      ...item,
+      enabled: !(item.enabled === false),
     }
-  }
-  return ''
+  },
 }
-// TODO(jc): end
+export const ZoneInsight = {
+  fromObject: (item?: PartialZoneInsight): ZoneInsight | undefined => {
+    // if item isn't set don't even try augmenting things
+    return isSet<PartialZoneInsight>(item)
+      ? ((item) => {
+        // figure out the connectedSubscription by looking at the connectTime
+        // and disconnectTime of the last subscription
+        const connectedSubscription: KDSSubscription | undefined = !Array.isArray(item.subscriptions)
+          ? undefined
+          : [item.subscriptions[item.subscriptions.length - 1]].find((item) => item.connectTime?.length && !item.disconnectTime)
+
+        // using the connectedSubscription find the config for the zone if it exists, is valid JSON and is not null and
+        // turn it into an object
+        const config: Record<string, unknown> = (() => {
+          const str = isSet<string>(connectedSubscription?.config) ? connectedSubscription.config : '{}'
```

Comment:
**Comment**: I noticed an issue with this utility. Take the following reduced example:

```ts
const array: { prop?: string }[] = []
const element = array.at(0)
const test = isSet<string>(element?.prop) ? element.prop : null
//                                          ^^^^^^^ 'element' is possibly 'undefined'.
```

When the TypeScript compiler infers that `element` has a union type like `Interface | undefined`, the `isSet` utility doesn’t function as a complete type guard anymore.

Note that this just happens to not be surfaced when using `array[0]` because by default, TypeScript considers `array[i]` safe and returns `T`. With `noUncheckedIndexedAccess` enabled, it would be `T | undefined` and the same error would happen, I think.

Reply:
🤔 

`str` is always a `string` here, I believe?
Also, if `connectedSubscription?.config` is a string, then we access it as `connectedSubscription.config` because we know its set?

Could you explain a little more what the issue is here, or is this just a comment?

Reply:
VS Code shows it. Not sure how to reproduce it another way.